apps/api/src/services/perps-data-processor.service.ts

```diff
@@ -749,6 +767,150 @@ export class PerpsDataProcessor {
+
+    let consecutiveFailures = 0;
+    let systemicFailureDetected = false;
+
+    for (let i = 0; i < successfulAgents.length; i += BATCH_SIZE) {
+      const batch = successfulAgents.slice(i, i + BATCH_SIZE);
+
+      // Use Promise.allSettled so one failure doesn't block others
+      const batchResults = await Promise.allSettled(
+        batch.map((agent) =>
+          calmarService.calculateAndSaveCalmarRatio(
+            agent.agentId,
+            competitionId,
+          ),
+        ),
+      );
+
+      // Count successes and failures in this batch
+      let batchFailures = 0;
+
+      batchResults.forEach((result, idx) => {
+        if (result.status === "fulfilled") {
+          results.successful++;
+        } else {
+          results.failed++;
+          batchFailures++;
+
+          const agent = batch[idx];
+          if (!agent) {
+            serviceLogger.error(
+              `[PerpsDataProcessor] Batch index mismatch at index ${idx}`,
+            );
+            return;
+          }
+          const errorMsg = `Agent ${agent.agentId}: ${result.reason?.message || "Unknown error"}`;
+
+          // Only store first N errors to prevent memory issues
+          if (results.errors.length < MAX_ERRORS_TO_STORE) {
+            results.errors.push(errorMsg);
+          } else if (results.errors.length === MAX_ERRORS_TO_STORE) {
+            results.errors.push(
+              `... and ${successfulAgents.length - i} more errors`,
+            );
+          }
+
+          // Log individual failures for debugging
+          serviceLogger.warn(
+            `[PerpsDataProcessor] Failed to calculate Calmar for ${agent?.agentId}: ${result.reason}`,
+          );
+        }
+      });
+
+      // Circuit breaker: Check if entire batch failed
+      const batchFailureRate = batchFailures / batch.length;
+      if (batchFailureRate === 1) {
+        consecutiveFailures++;
+
+        if (consecutiveFailures >= MAX_CONSECUTIVE_FAILURES) {
+          systemicFailureDetected = true;
+          const remainingAgents = successfulAgents.length - (i + BATCH_SIZE);
+
+          serviceLogger.error(
+            `[PerpsDataProcessor] Circuit breaker triggered: ${consecutiveFailures} consecutive batch failures. ` +
+              `Stopping Calmar calculations. ${remainingAgents} agents skipped.`,
+          );
+
+          // Mark remaining agents as failed
+          results.failed += remainingAgents;
+
+          // Add circuit breaker message to errors
+          if (results.errors.length < MAX_ERRORS_TO_STORE) {
+            results.errors.push(
+              `Circuit breaker triggered after ${consecutiveFailures} consecutive batch failures. ` +
+                `${remainingAgents} agents skipped.`,
+            );
+          }
+
+          break; // Exit the loop
+        }
+      } else {
+        // Reset counter if batch had any successes
+        consecutiveFailures = 0;
```

Comment:
Processing will continue even through back-to-back batches with a 90% failure rate. Do we want the circuit break to work this way?